### DISCUSSION TITLE
Crop to PSD clip path, maintaining transparency

### DISCUSSION THREAD

**Discussion by digitracey** [ImageMagick/ImageMagick#7679] [Help] [answered]
### ImageMagick version

 7.1.1-33

### Operating system, version and so on

Windows

### Description

Please note that I am working with a DAM system that only accepts a single-line CMD in the configuration settings and cannot use intermediary files, so I need to achieve this in one line without generating other files.

I need to crop this PSD (https://keyshot-my.sharepoint.com/:i:/p/tracey_spencer/EarTYX_jsZxMsWRUlqZS2-wBTX_wDHbM6cbvbqzMsKaBpg?e=PeOUNy) to the 'Square' clip path, maintaining the padding and transparency and convert it to WebP.

The closest I can get is this, but it gives it a white background instead of maintaining the transparency:
 `magick RenditionI.psd -layers merge -alpha transparent -clip-path Square -alpha opaque +repage -trim RenditionI.webP`

https://keyshot-my.sharepoint.com/:u:/p/tracey_spencer/EYo7IInRDp1FgIWyU0QQxIABllCA-ULN1dTitA7UflgM7w?e=eEyet2

Ideally, I need the background to be transparent, alternatively, I need to set it to another colour. I have tried removing white, but it also removes it from within the watch.

I am using ImageMagick 7.1.1-33

Sorry about the links, I wasn't allowed to upload PSDs or WebPs; they are unsupported file types here.

**Comment by snibgo** [comment]
What images are in the file?
```
magick RenditionI.psd info:
RenditionI.psd[0] PSD 5118x7323 5118x7323+0+0 8-bit sRGB 110.344MiB 0.001u 0:00.000
RenditionI.psd[1] PSD 2642x675 2642x675+1277+6372 8-bit sRGB 0.002u 0:00.001
RenditionI.psd[2] PSD 3327x721 3327x721+925+6337 8-bit sRGB 0.003u 0:00.002
RenditionI.psd[3] PSD 4273x6298 4273x6298+565+515 8-bit sRGB 0.004u 0:00.003
```
IM can see 4 images in the PSD file. I assume you want the first, so "[0]".

"magick RenditionI.psd[0] -verbose info:" tells us that IM can't see a name for the clip path.

We can see the mask with this:
```
magick RenditionI.psd[0] -format %[8BIM:1999,2998:#1] info: | %IMG7%magick - clip.png
```
clip.png is black where the watch is, and white elsewhere (including the background and the strap.

> I need to crop this PSD (...) to the 'Square' clip path, maintaining the padding and transparency.

I'm not sure what you mean by "maintaining the padding and transparency". What "padding"? As far as I can see, the input image has no transparency within the mask. So I think you want to replace the input's transparency by the mask.

So:
```
magick RenditionI.psd[0] -format %[8BIM:1999,2998:#1] info: | magick RenditionI.psd[0] ( - -channel RGB -negate +channel ) -alpha off -compose CopyOpacity -composite -trim +repage out.webp
```
That command creates an antialiased mask. A simpler alternative creates an aliased mask, so the mask is binary, and pixels are either fully transparent or fully opaque:
```
magick RenditionI.psd[0] -alpha transparent -clip -alpha opaque +clip -trim +repage out3.webp
```

**Reply by digitracey** [reply]
Thanks for your reply. I should have explained that there are two clip paths; I need to use the one called Square.
<img width="218" alt="image" src="https://github.com/user-attachments/assets/4ad3cbad-dc9d-4247-a5af-f5ca2dccc4c7">

If you select the clip path in Photoshop, you will see that it is a Sqaure around part of the watch.

<img width="336" alt="image" src="https://github.com/user-attachments/assets/3eeb4d70-bb03-41f7-9baa-090e7346e9c4">
 
The padding between the watch and the edge of the clip path is what I was referring to when I said I need to retain the padding.

So, the final image should look like this (without the border):
<img width="325" alt="image" src="https://github.com/user-attachments/assets/044c8d59-44a9-4eda-8cc6-1f715e4c1d7c">

What I am getting with `magick RenditionI.psd -layers merge -alpha transparent -clip-path Square -alpha opaque +repage -trim RenditionI.webP` is this (without the border):
<img width="429" alt="image" src="https://github.com/user-attachments/assets/91a6fcd9-3ab4-4ff4-9eba-29f6a926ad13">

I have to put this into the rendition configuration settings of a Digital Asset Management (DAM) system, which can only accept single lines, so I cannot use the first example you gave me. However, I tried it locally and it is using the wrong clip path. When I run it locally using the correct clip path `magick RenditionI.psd[0] -format %[8BIM:1999,2998:#2] info: | magick RenditionI.psd[0] ( - -channel RGB -negate +channel ) -alpha off -compose CopyOpacity -composite -trim +repage out.webp` it gives the same as above, the correct part of the image, but with a white background instead of transparent.

Your second example, I can use in the DAM system, but again it is using the wrong clip path, when I change it to `magick RenditionI.psd[0] -alpha transparent -clip-path Square -alpha opaque +clip -trim +repage out3.webp` I get the same as above, the correct part of the image, but with a white background instead of transparent.

**Comment by snibgo** [comment]
With the square path, we need to use the minimum of two alphas: the original alpha, and the alpha from the clip path.

```
magick RenditionI.psd[0] ( +clone -alpha extract +write mpr:ALP1 +delete ) -alpha transparent -clip-path Square -alpha opaque +clip ( +clone -alpha extract mpr:ALP1 -fx "min(u,v)" ) -alpha off -compose CopyOpacity -composite -trim +repage out4.webp
```

**Comment by digitracey** [comment]
Thanks. That's almost perfect. It keeps the correct part of the watch and the transparency, but not the padding, so it is not square, but I should be able to play around with adding padding to get it right.

<img width="367" alt="image" src="https://github.com/user-attachments/assets/58e22587-ea97-4ac1-877f-28825ddee05c">

Thank you very much for your help.

**Reply by digitracey** [reply]
I spoke too soon. When I try to add padding (using extent), I lose the content.

<img width="488" alt="image" src="https://github.com/user-attachments/assets/00c50527-1b1b-48ee-8f81-010bafdbc311">

**Reply by fmw42** [reply]
You do not show your command line. Please do so.

**Reply by fmw42** [reply]
Mac OSX Ventura, IM 7.1.1.39. 

When I try to run your command or a lesser command, I get warnings and Inkscape errors. It seems to be having trouble accessing the Square clip-path. If I remove the -clip-path, the following runs fine

`magick RenditionI.psd[0] \( +clone -alpha extract +write mpr:ALP1 +delete \) -alpha transparent -clip-path Square null:`

** (<unknown>:1496): WARNING **: 10:05:17.214: Can't open file: /tmp/magick-5B72wyci12Shdf90v3Vu-PK4KxONgvNJ (doesn't exist)

** (<unknown>:1496): WARNING **: 10:05:17.215: Could not detect file format. Tried to open it as an SVG anyway but this also failed.

** (<unknown>:1496): WARNING **: 10:05:17.215: Can't open file: /tmp/magick-5B72wyci12Shdf90v3Vu-PK4KxONgvNJ (doesn't exist)
ink_file_open: '/tmp/magick-5B72wyci12Shdf90v3Vu-PK4KxONgvNJ' cannot be opened!
InkscapeApplication::document_open: Failed to open: /tmp/magick-5B72wyci12Shdf90v3Vu-PK4KxONgvNJ
ConcreteInkscapeApplication::on_open: failed to create document!

**Comment by fmw42** [comment]
Even though I get those warnings and errors, your command seems to work for me on IM 7.1.1.39 on Mac OSX Ventura

`magick RenditionI.psd[0] \( +clone -alpha extract +write mpr:ALP1 +delete \) -alpha transparent -clip-path Square -alpha opaque +clip \( +clone -alpha extract mpr:ALP1 -fx "min(u,v)" \) -alpha off -compose CopyOpacity -composite -trim +repage -gravity center out4.webp`

[out4.webp.zip](https://github.com/user-attachments/files/17314045/out4.webp.zip)

**Comment by snibgo** [comment, marked as answer]
> I spoke too soon. When I try to add padding (using extent), I lose the content.

You don't show your command, so we can only guess at the problem.

This works for me:
```
magick 
  RenditionI.psd[0] 
  ( +clone -alpha extract +write mpr:ALP1 +delete ) 
  -alpha transparent -clip-path Square -alpha opaque +clip 
  ( +clone -alpha extract mpr:ALP1 -fx "min(u,v)" ) 
  -alpha off -compose CopyOpacity -composite 
  -trim +repage 
  -background None -compose Over -gravity Center 
  -extent %[fx:max(w,h)]x%[fx:max(w,h)] 
  x.png
```
For readability, I have split it into lines. Either join them up again, or use line-continuation characters at the end of each line.

**Comment by digitracey** [comment]
`magick RenditionI.psd[0] ( +clone -alpha extract +write mpr:ALP1 +delete ) -alpha transparent -clip-path Square -alpha opaque +clip ( +clone -alpha extract mpr:ALP1 -fx "min(u,v)" ) -alpha off -compose CopyOpacity -composite -trim +repage -background None -compose Over -gravity Center -extent %[fx:max(w,h)]x%[fx:max(w,h)] out10.webp` worked great, thanks!
